File: engine/events.py
# Battle events for logging / replay

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class SwitchEvent:
    player: int  # 1 or 2
    from_name: str | None
    to_name: str

@dataclass
class MoveEvent:
    player: int
    pokemon_name: str
    move_name: str
    damage: int
    effectiveness: float
    is_crit: bool
    target_hp_remaining: int

@dataclass
class FaintEvent:
    player: int
    pokemon_name: str

@dataclass
class StruggleEvent:
    player: int
    pokemon_name: str
    damage: int

@dataclass
class MissEvent:
    player: int
    pokemon_name: str
    move_name: str

@dataclass
class StatusMoveEvent:
    """Status move used -- no-op in v1."""
    player: int
    pokemon_name: str
    move_name: str

@dataclass
class StatusAppliedEvent:
    player: int
    pokemon_name: str
    status: str

@dataclass
class StatusCuredEvent:
    player: int
    pokemon_name: str
    status: str

@dataclass
class StatusPreventedEvent:
    player: int
    pokemon_name: str
    status: str
    reason: str  # "fully paralyzed", "fast asleep", "frozen solid"

@dataclass
class ResidualDamageEvent:
    player: int
    pokemon_name: str
    status: str
    damage: int

@dataclass
class ConfusionAppliedEvent:
    player: int
    pokemon_name: str

@dataclass
class ConfusionHitSelfEvent:
    player: int
    pokemon_name: str
    damage: int

@dataclass
class StatChangeEvent:
    player: int
    pokemon_name: str
    stat: str
    stages: int  # positive = boost, negative = drop

@dataclass
class HealEvent:
    player: int
    pokemon_name: str
    amount: int
    source: str  # "drain", "recover", etc.

@dataclass
class FlinchEvent:
    player: int
    pokemon_name: str

@dataclass
class SpikesSetEvent:
    player: int  # side that received spikes

@dataclass
class SpikesDamageEvent:
    player: int
    pokemon_name: str
    damage: int

@dataclass
class ScreenSetEvent:
    player: int
    screen: str  # "reflect" or "light_screen"

@dataclass
class ScreenExpiredEvent:
    player: int
    screen: str

@dataclass
class ProtectEvent:
    player: int
    pokemon_name: str
    success: bool

@dataclass
class LeechSeedAppliedEvent:
    player: int  # the seeded pokemon's side
    pokemon_name: str

@dataclass
class LeechSeedDrainEvent:
    player: int  # side being drained
    pokemon_name: str
    damage: int

@dataclass
class PhazeEvent:
    player: int  # side forced to switch
    pokemon_name: str
    forced_in: str

@dataclass
class HazeEvent:
    player: int

@dataclass
class WeatherSetEvent:
    player: int
    weather: str  # "sun", "rain", "sandstorm"

@dataclass
class WeatherDamageEvent:
    player: int
    pokemon_name: str
    damage: int

@dataclass
class WeatherExpiredEvent:
    weather: str


Event = (
    SwitchEvent | MoveEvent | FaintEvent | StruggleEvent | MissEvent
    | StatusMoveEvent | StatusAppliedEvent | StatusCuredEvent
    | StatusPreventedEvent | ResidualDamageEvent
    | ConfusionAppliedEvent | ConfusionHitSelfEvent
    | StatChangeEvent | HealEvent | FlinchEvent
    | SpikesSetEvent | SpikesDamageEvent
    | ScreenSetEvent | ScreenExpiredEvent
    | ProtectEvent | LeechSeedAppliedEvent | LeechSeedDrainEvent
    | PhazeEvent | HazeEvent
    | WeatherSetEvent | WeatherDamageEvent | WeatherExpiredEvent
)

File: crystal_engine/src/events.rs
// Battle event log (replay/observation layer)

#[derive(Debug, Clone)]
pub enum Event {
    Switch {
        player: u8,
        from_name: Option<String>,
        to_name: String,
    },
    Move {
        player: u8,
        pokemon_name: String,
        move_name: String,
        damage: u16,
        effectiveness: f32,
        is_crit: bool,
        target_hp_remaining: u16,
    },
    Faint {
        player: u8,
        pokemon_name: String,
    },
    Struggle {
        player: u8,
        pokemon_name: String,
        damage: u16,
    },
    Miss {
        player: u8,
        pokemon_name: String,
        move_name: String,
    },
    StatusMove {
        player: u8,
        pokemon_name: String,
        move_name: String,
    },
    StatusApplied {
        player: u8,
        pokemon_name: String,
        status: String,
    },
    StatusCured {
        player: u8,
        pokemon_name: String,
        status: String,
    },
    StatusPrevented {
        player: u8,
        pokemon_name: String,
        status: String,
        reason: String,
    },
    ResidualDamage {
        player: u8,
        pokemon_name: String,
        status: String,
        damage: u16,
    },
    ConfusionApplied {
        player: u8,
        pokemon_name: String,
    },
    ConfusionHitSelf {
        player: u8,
        pokemon_name: String,
        damage: u16,
    },
    StatChange {
        player: u8,
        pokemon_name: String,
        stat: String,
        stages: i8,
    },
    Heal {
        player: u8,
        pokemon_name: String,
        amount: u16,
        source: String,
    },
    Flinch {
        player: u8,
        pokemon_name: String,
    },
    SpikesSet {
        player: u8,
    },
    SpikesDamage {
        player: u8,
        pokemon_name: String,
        damage: u16,
    },
    ScreenSet {
        player: u8,
        screen: String,
    },
    ScreenExpired {
        player: u8,
        screen: String,
    },
    Protect {
        player: u8,
        pokemon_name: String,
        success: bool,
    },
    LeechSeedApplied {
        player: u8,
        pokemon_name: String,
    },
    LeechSeedDrain {
        player: u8,
        pokemon_name: String,
        damage: u16,
    },
    Phaze {
        player: u8,
        pokemon_name: String,
        forced_in: String,
    },
    Haze {
        player: u8,
    },
    WeatherSet {
        player: u8,
        weather: String,
    },
    WeatherDamage {
        player: u8,
        pokemon_name: String,
        damage: u16,
    },
    WeatherExpired {
        weather: String,
    },
}
